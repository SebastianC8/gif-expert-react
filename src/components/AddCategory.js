import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { capitalizeString } from '../helpers/capitalize';

/**
 * Se recibe una FUNCIÓN en los props
 */
export const AddCategory = ({ setCategories }) => {

    /**
     * Hook para guardar nueva categoría
     */
    const [inputValue, setInputValue] = useState('');

    /**
     * Evento change para capturar el valor del input
     */
    const handleInputChange = (e) => {
        setInputValue(capitalizeString(e.target.value));
    }

    /**
     * Evento submit para guardar nueva categoría
     * 
     * Se utiliza función hook setCategories() de componente <GifExpertApp/> para añadir nueva categoría
     */
    const handleSubmit = (e) => {

        e.preventDefault();
        
        if (inputValue) {
            setCategories((category) => [inputValue, ...category]);
            setInputValue('');
        }
        
    }

    return (
        <form onSubmit = { handleSubmit }>
            <input type="text" value = { inputValue } onChange = { handleInputChange }/>
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}