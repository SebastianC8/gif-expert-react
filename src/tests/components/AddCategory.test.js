import React from "react";
import { shallow } from "enzyme";
import { AddCategory } from "../../components/AddCategory";

describe('Componente <AddCategory/>', () => {

    /**
     * Definición de función
     */
    const setCategories = jest.fn();

    /**
     * Componente
     */
    let wrapper;

    /**
     * Limpiar llamado a la instancia
     */
    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories = { setCategories }/>);
    });
    
    /**
     * Snapshot ok
     */
    test('Mostrar componente ok', () => {
        
        expect(wrapper).toMatchSnapshot();

    });

    /**
     * Evento change en input de búsqueda
     */
    test('Cambios en input', () => {
        
        const input = wrapper.find('input');
        const value = 'Valor input';
        input.simulate('change', { target: { value } });

        expect(wrapper.find('p').text().trim()).toBe(value);

    });

    /**
     * Valida el valor del input y como no se le está mandando,
     * no debe llamarse la función setCategories()
     */
    test('NO debe de ejecutarse', () => {
        
        wrapper.find('form').simulate('submit', { preventDefault(){} });
        expect(setCategories).not.toHaveBeenCalled();

    });

    /**
     * Valida que llame la función setCategories() y limpie el input
     */
    test('Debe llamar setCategories() y limpiar input', () => {
    
        const value = 'Text input';

        wrapper.find('input').simulate('change', { target: { value } });
        wrapper.find('form').simulate('submit', { preventDefault(){} });

        expect(setCategories).toHaveBeenCalled();
        expect(wrapper.find('input').prop('value')).toBe('');

    });

});
