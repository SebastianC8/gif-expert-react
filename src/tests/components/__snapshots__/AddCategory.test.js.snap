// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componente <AddCategory/> Mostrar componente ok 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
